### PR TITLE
[3.7] bpo-36389: _PyObject_IsFreed() now also detects uninitialized memory (GH-12770)

### PR DESCRIPTION
**Pull request by vstinner**
* [bpo-36389](https://bugs.python.org/issue36389): _PyObject_IsFreed() now also detects uninitialized memory (GH-12770)
* [bpo-36389](https://bugs.python.org/issue36389): Change PyMem_SetupDebugHooks() constants (GH-12782)

https://bugs.python.org/issue36389

### CODE REVIEW

File: Include/pymem.h [hunk 1]
```diff
@@ -55,8 +55,6 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
 PyAPI_FUNC(PyObject*) _PyTraceMalloc_GetTraceback(
     unsigned int domain,
     uintptr_t ptr);
-
-PyAPI_FUNC(int) _PyMem_IsFreed(void *ptr, size_t size);
```
Comment (asottile) [inline Include/pymem.h:59]: (it's too late now) but should this have been backported? removing this symbol is ~technically an abi break (despite it being a "private" function)

packaging of 3.7.4 is complaining about disappearing symbols :(

Comment (vstinner) [reply Include/pymem.h:59]: We don't provide any warranty on private symbols (prefixed by _Py or _PY): they can disappear anytime. The main issue was that we exposed it in the first place :-)

How do you check for symbols? My notes on the Python "stable" ABI: https://pythoncapi.readthedocs.io/stable_abi.html

**Comment by asottile** [comment]
> We don't provide any warranty on private symbols

Doesn't surprise :)

> How do you check for symbols?

debian lists the exported symbols of the `.so` file using [dh_makeshlibs](https://manpages.debian.org/testing/debhelper/dh_makeshlibs.1.en.html) (the reasoning there being as packages upgrade and others are built against them, they can accurately represent the `>=` relationship with the absolute minimum version that's necessary -- instead of unnecessarily pinning to the latest version)

(it's actually quite annoying and probably the most time-consuming part of packaging python for me)

Here's (for example) an "update symbols" commit: https://github.com/deadsnakes/python3.7/commit/c5aec48a264af853c3abf156b352fba4c77a3a8c

(or one with a more involved set of changes): https://github.com/deadsnakes/python3.7/commit/55bead25d3bed3c87b7b16914c15465405f74950

**Comment by vstinner** [comment]
Maybe I should modify internal headers to not export symbols like _Py_ResetForceASCII() which should not be used outside Python. It is only declared in internal headers: Include/internal/pycore_fileutils.h. It is declared with PyAPI_FUNC(). Maybe I should replace PyAPI_FUNC() with extern.

Do you *have* to list private symbols starting with _Py?

**Comment by asottile** [comment]
> Do you have to list private symbols starting with _Py?

I'm not sure debian has a concept of "private by convention" (it might! I just don't know) -- to be honest I don't actually know too much about the makeshlibs utility but I'll do some research :+1:

**Pull request merged**